### PR TITLE
#6547 add continious deletion of IOU amount on long press

### PR DESCRIPTION
**Pull request by dklymenk**
### Details

This PR adds a long press handler to delete button on mobile numpad that starts continuous deletion of characters until it is released. 

### Fixed Issues

$ https://github.com/Expensify/App/issues/6547

### Tests / QA Steps

1. Navigate to a conversation
2. Start the Request money flow
3. Type some amount with several digits
4. Press and hold the delete button (or Backspace key on web and desktop)
5. The amount should be continuously deleted until the key is released

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Mobile Web

https://user-images.githubusercontent.com/64093836/145180245-2802cd4d-7001-425d-bd2b-4dc89089d8f0.mp4

#### iOS

https://user-images.githubusercontent.com/64093836/145092374-15bc7c35-e057-4aac-98e4-2fa942a3d78d.mp4

#### Android

https://user-images.githubusercontent.com/64093836/145092403-6ba2dc97-5f9f-4cc4-852b-d82f763321c2.mp4

### CODE REVIEW

File: src/components/BigNumberPad.js [hunk 1]
```diff
@@ -41,5 +54,6 @@ const BigNumberPad = props => (
 
 BigNumberPad.propTypes = propTypes;
 BigNumberPad.displayName = 'BigNumberPad';
+BigNumberPad.defaultProps = defaultProps;
```
Comment (parasharrajat) [inline src/components/BigNumberPad.js:57]:
```suggestion
BigNumberPad.defaultProps = defaultProps;
BigNumberPad.displayName = 'BigNumberPad';
```

File: src/components/ExpensifyButton.js [hunk 2]
```diff
@@ -158,6 +167,12 @@ class ExpensifyButton extends Component {
         return (
             <Pressable
                 onPress={this.props.onPress}
+                onLongPress={this.props.onLongPress}
+                onPressIn={() => ControlSelection.block()}
+                onPressOut={() => {
+                    this.props.onPressOut();
+                    ControlSelection.unblock();
+                }}
```
Comment (parasharrajat) [inline src/components/ExpensifyButton.js:175]: Not a good idea. Just move selection blocking logic to the BigNumberPad. We should not extend the scope of this beyond what is needed for this issue.

File: src/pages/iou/steps/IOUAmountPage.js [hunk 3]
```diff
@@ -155,6 +157,33 @@ class IOUAmountPage extends React.Component {
         });
     }
 
+    /**
+     * Handle long press key on number pad.
+     * Only handles the '<' key and starts the continious input deletion timer
```
Comment (parasharrajat) [inline src/pages/iou/steps/IOUAmountPage.js:162]:
```suggestion
     * Only handles the '<' key and starts the continuous input timer.
```

File: src/pages/iou/steps/IOUAmountPage.js [hunk 4]
```diff
@@ -155,6 +157,33 @@ class IOUAmountPage extends React.Component {
...
+            }));
+        }, 100);
+
+        this.setState({timer});
+    }
+
+    /**
+     * Clear the timer for continious input deletion
+     */
+    clearDeletionTimer() {
```
Comment (parasharrajat) [inline src/pages/iou/steps/IOUAmountPage.js:183]:
```suggestion
    clearLongPressTimer() {
```

File: src/pages/iou/steps/IOUAmountPage.js [hunk 5]
```diff
@@ -155,6 +157,33 @@ class IOUAmountPage extends React.Component {
...
+            this.setState(prevState => ({
+                amount: prevState.amount.slice(0, -1),
+            }));
+        }, 100);
+
+        this.setState({timer});
+    }
+
+    /**
+     * Clear the timer for continious input deletion
```
Comment (parasharrajat) [inline src/pages/iou/steps/IOUAmountPage.js:181]:
```suggestion
     * Clear the timer for continuous input.
```

File: src/pages/iou/steps/IOUAmountPage.js [hunk 6]
```diff
@@ -155,6 +157,33 @@ class IOUAmountPage extends React.Component {
...
     }
 
+    /**
+     * Handle long press key on number pad.
+     * Only handles the '<' key and starts the continious input deletion timer
+     *
+     * @param {String} key
+     */
+    handleLongPressNumberPad(key) {
+        if (key !== '<') {
```
Comment (parasharrajat) [inline src/pages/iou/steps/IOUAmountPage.js:167]: Better add a comment here saying that only allow deleting for now.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Comment by dklymenk** [comment]
Applied the changes suggested by @parasharrajat

**Review by parasharrajat — DISMISSED** [review]
LGTM and tests well. 

All yours @iwiznia

File: src/pages/iou/steps/IOUAmountPage.js [hunk 7]
```diff
@@ -155,6 +157,34 @@ class IOUAmountPage extends React.Component {
         });
     }
 
+    /**
+     * Handle long press key on number pad.
+     * Only handles the '<' key and starts the continuous input timer.
+     *
+     * @param {String} key
+     */
+    handleLongPressNumberPad(key) {
```
Comment (iwiznia) [inline src/pages/iou/steps/IOUAmountPage.js:166]: Hmmm shouldn't all this logic be on the BigNumberPad component? Do we expect any BigNumberPad were we don't want long presses to work? I would think we always want this behavior, so reimplementing it on each caller sounds worse than just incorporating it to that component.

Comment (parasharrajat) [reply src/pages/iou/steps/IOUAmountPage.js:166]: Good point, I was thinking the same but had doubts.

Comment (parasharrajat) [reply src/pages/iou/steps/IOUAmountPage.js:166]: Basically, my point was Continues keypress event timer logic should be inside the BigNumberPad. This callback should just be doing side-effect.

Comment (dklymenk) [reply src/pages/iou/steps/IOUAmountPage.js:166]: But that component doesn't have any state itself, it is simply responsible to passing the callbacks to the buttons.

Or do you think we should pass some 'deleteSingleChar' prop to the number pad and call it with `setInterval` inside the `BigNumberPad`?

Comment (parasharrajat) [reply src/pages/iou/steps/IOUAmountPage.js:166]: Yup, so you just need to trigger the passed callback as you do now. But longPress will trigger that multiple times.

Comment (dklymenk) [reply src/pages/iou/steps/IOUAmountPage.js:166]: Ah, I got you. There is no need for a new prop, I need to call the one that is already used for single press. :+1:

Comment (dklymenk) [reply src/pages/iou/steps/IOUAmountPage.js]: @parasharrajat @iwiznia 
I have updated the PR with the change you requested, please review.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Do not follow StyleGuide. Please have a read of https://github.com/Expensify/App/blob/main/STYLING.md and https://github.com/Expensify/App/blob/main/STYLE.md.

**Comment by dklymenk** [comment]
> Do not follow StyleGuide. Please have a read of https://github.com/Expensify/App/blob/main/STYLING.md and https://github.com/Expensify/App/blob/main/STYLE.md.

Do you want me to rewrite the component as class component? Before committing I have checked for usages of `useState` in the app and found [at least one component](https://github.com/Expensify/App/blob/5f154bcd1e6bfbe4a6636c4e04d7303db551b660/src/components/AddressSearch.js#L43) that uses it, so it didn't seem like I am breaking the rules.

Anyway, just let me know if that's what you meant. I am slightly confused by you referencing the styling doc where this PR doesn't touch any styles whatsoever.

Thanks.

**Comment by parasharrajat** [comment]
> Do you want me to rewrite the component as class component? Before committing I have checked for usages of useState in the app and found at least one component that uses it, so it didn't seem like I am breaking the rules.

Yeah, this should be converted to the class component. You should follow styleGuide as much as possible. Exceptions could be made but do you think that is necessary here. I am sure you read the comment above https://github.com/Expensify/App/blob/5f154bcd1e6bfbe4a6636c4e04d7303db551b660/src/components/AddressSearch.js#L38.

**Comment by dklymenk** [comment]
Ah, no. I missed that comment, sorry. I will try to update the `BigNumberPad` to use `React.Component`.

**Comment by dklymenk** [comment]
Updated PR with requested change: rewrite `BigNumberPad` as class component. Please review. @parasharrajat @iwiznia

File: src/components/BigNumberPad.js [hunk 8]
```diff
@@ -17,27 +18,64 @@ const padNumbers = [
...
-                            text={column}
-                            onPress={() => props.numberPressed(column)}
-                        />
-                    );
-                })}
+class BigNumberPad extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {timer: null};
```
Comment (parasharrajat) [inline src/components/BigNumberPad.js:25]:
```suggestion
        this.state = {
        	timer: null,
        };
```

File: src/components/BigNumberPad.js [hunk 9]
```diff
@@ -17,27 +18,64 @@ const padNumbers = [
...
+     * Handle long press key on number pad.
+     * Only handles the '<' key and starts the continuous input timer.
+     *
+     * @param {String} key
+     */
+    handleLongPress(key) {
+        // Only handles deleting.
+        if (key !== '<') {
+            return;
+        }
```
Comment (parasharrajat) [inline src/components/BigNumberPad.js:38]: This should be part of BigNumberPad. This should be handled in the parent component. The only job of this method should be to call the long press callback on time intervals emitting continuous long-press.

File: src/components/BigNumberPad.js [hunk 10]
```diff
@@ -17,27 +18,64 @@ const padNumbers = [
...
+        this.setState(
+            {
+                timer: setInterval(() => {
+                    this.props.numberPressed(key);
+                }, 100),
```
Comment (parasharrajat) [inline src/components/BigNumberPad.js:43]: Not very intuitive.
Better set the interval above and then set the handle to State in next line.

File: src/components/BigNumberPad.js [hunk 11]
```diff
@@ -17,27 +18,64 @@ const padNumbers = [
...
-        ))}
-    </View>
-);
+        );
+    }
+}
+
 
 BigNumberPad.propTypes = propTypes;
 BigNumberPad.displayName = 'BigNumberPad';
```
Comment (parasharrajat) [inline src/components/BigNumberPad.js:81]:
```suggestion
```

File: src/components/BigNumberPad.js [hunk 12]
```diff
@@ -17,27 +18,64 @@ const padNumbers = [
...
+                                    key={column}
+                                    style={[styles.flex1, marginLeft]}
+                                    text={column}
+                                    onLongPress={() => this.handleLongPress(column)}
+                                    onPress={() => this.props.numberPressed(column)}
+                                    onPressOut={() => {
+                                        clearInterval(this.state.timer);
+                                        ControlSelection.unblock();
+                                    }}
+                                    onPressIn={ControlSelection.block}
```
Comment (parasharrajat) [inline src/components/BigNumberPad.js:68]: NAB: will look better if you can move this before `onPressOut`.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/BigNumberPad.js (hunk 9 above)
Comment (dklymenk) [reply src/components/BigNumberPad.js:40]: Sorry, but I don't quite understand. There is no more long press callback in that context. I think we have agreed that this entire logic should be moved to BigNumberPad component.

I have just reread the thread above and @iwiznia clearly says "shouldn't all this logic be on the BigNumberPad component". If we are to do
```js
        if (key !== '<') {
            return;
        }
```
inside parent component we would need to also pass some new arg to callback to indicate that the keyPress was started by the timer.  And overall moving it to parent component would mean that we'll have timers on every button and not just the `<`.

Comment (iwiznia) [reply src/components/BigNumberPad.js:40]: I agree with @dklymenk, not sure why would anyone want to handle the long presses differently. But even if at some point we do want that, we can extract the logic then.

**Review by iwiznia — APPROVED** [review]

**Comment by iwiznia** [comment]
I think this solution is good, but will let @parasharrajat merge or keep engaging on the discussion above.

File: src/components/BigNumberPad.js (hunk 9 above)
Comment (parasharrajat) [reply src/components/BigNumberPad.js:40]: OK, I give it a go for now. I still feel that all keys should have the same behaviour irrespective of the Key. If we only want to handle long-press for `<` then it should be a parent to control that.

**Review by parasharrajat — APPROVED** [review]
LGTM. we are good to merge @iwiznia .

**Pull request merged**